src/Pos/Explorer/Web/Server.hs

```diff
@@ -126,62 +147,126 @@ getAddressSummary cAddr = cAddrToAddr cAddr >>= \addr -> case addr of
     PubKeyAddress sid _ -> do
         balance <- fromMaybe (mkCoin 0) <$> GS.getFtsStake sid
         -- TODO: add number of coins when it's implemented
-        txs <- allTxs
-        let transactions = map (toTxDetailed addr) $
-                           filter (any (\txOut -> txOutAddress txOut == addr) .
-                               _txOutputs . tiTx) txs
+        -- TODO: retrieve transactions from something like an index
+        let transactions = []
         return $ CAddressSummary cAddr 0 balance transactions
     _ -> throwM $
          Internal "Non-P2PKH addresses are not supported in Explorer yet"
 
+getTxSummary :: ExplorerMode m => CTxId -> m CTxSummary
+getTxSummary cTxId = do
+    -- There are two places whence we can fetch a transaction: MemPool and DB.
+    -- However, TxExtra should be added in the DB when a transaction is added
+    -- to MemPool. So we start with TxExtra and then figure out whence to fetch
+    -- the rest.
+    txId <- cTxIdToTxId cTxId
+    txExtraMaybe <- GS.getTxExtra txId
+
+    when (isNothing txExtraMaybe) $
+        throwM $ Internal "Transaction not found"
+
+    let txExtra = fromJust txExtraMaybe
+        blockchainPlace = teBlockchainPlace txExtra
+        inputOutputs = teInputOutputs txExtra
+
+    let convertTxOutputs = map (\txOut ->(toCAddress $ txOutAddress txOut, txOutValue txOut))
+
+    -- TODO: here and in getMempoolTxs/getBlockchainTxs we do two things wrongly:
+    -- 1. If the transaction is found in the MemPool, we return *starting
+    --    time of the current block* as the time when it was issued.
+    -- 2. If the transaction comes from the blockchain, we return *block
+    --    slot starting time* as the time when it was issued.
+    -- This needs to be fixed.
+    (ctsTxTimeIssued, ctsBlockTimeIssued, ctsBlockHeight, ctsOutputs) <-
+        case blockchainPlace of
+            Nothing -> do
+                -- Fetching transaction from MemPool.
+                ts <- currentTimeSlotting
+                tx <- fetchTxFromMempoolOrFail txId
+                let txOutputs = convertTxOutputs $ _txOutputs tx
+                pure (Just (toPosixTime ts), Nothing, Nothing, txOutputs)
+            Just (headerHash, txIndexInBlock) -> do
+                -- Fetching transaction from DB.
+                maybeBlock <- DB.getBlock @SscGodTossing headerHash
+                when (isNothing maybeBlock) $
+                    throwM $ Internal "TxExtra says tx is in nonexistent block"
+                let block = fromJust maybeBlock
+                case block of
+                    Left _ -> throwM $ Internal "TxExtra says tx is in genesis block"
```

Comment:
@akegalj I've tried it, and it doesn't look more readable, since the `where` block is far away from this code, and it's inconvenient to scroll down and up again. What do you think?